arch/mips/core/thread.c

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2017 Imagination Technologies Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <arch/cpu.h>
+#include <kernel_structs.h>
+#include <wait_q.h>
+#include <string.h>
+#include <soc.h>
+
+void _thread_entry_wrapper(k_thread_entry_t thread,
+			   void *arg1,
+			   void *arg2,
+			   void *arg3);
+
+void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
+		 size_t stack_size, k_thread_entry_t entry,
+		 void *arg1, void *arg2, void *arg3,
+		 int prio, unsigned int options)
+{
+	_ASSERT_VALID_PRIO(prio, entry);
+
+	char *stack_memory = K_THREAD_STACK_BUFFER(stack);
+	char *ctx = (char *)STACK_ROUND_DOWN(stack + stack_size -
+		sizeof(struct gpctx));
+
+	thread->callee_saved.sp = (u32_t)ctx;
+
+	struct gpctx stack_init = {
```

Comment:
i don't like the memcpy this requires, it seems unnecessary.
Why not cast &stack_init to a struct gpctx * and initialize directly, after the call to _new_thread_init()?

Reply:
I agree, but would that not be a strict aliasing violation in cases where the stack is not made up of chars?

Reply:
not sure what you mean here.
K_THREAD_STACK_BUFFER() return type is always char *.

Reply:
I assumed you meant `ctx` should be cast to `struct gpctx *` and then initialised (as `&stack_init` is already a `struct gpctx *`). If so, I think that violates strict aliasing b/c `ctx` is just an offset from `stack` which is of the potentially incompatible type `k_thread_stack_t`. I am not calling `K_THREAD_STACK_BUFFER()` to initialise `ctx`, but even if I was, I think my point still stands (but I could be totally wrong)...

Reply:
Your code is wrong above, You should get the underlying buffer of the stacl with K_THREAD_STACK_BUFFER()

Reply:
have a look at the x86 port of _new_thread() to see what I am looking for.

Reply:
You're right, I will change that. You haven't quite convinced me to get rid of the `memcpy` yet though!

Reply:
K_THREAD_STACK_BUFFER() returns a char *. Do the appropriate math on it and cast your data structure. Why is this a problem? Please, get rid of memcpy(). It casts its arguments to char* anyway. If we are unable to treat stack buffers at character buffers the whole OS won't work.

Reply:
I'm not trying to be difficult here. If it is guaranteed that the stack buffer will always be a char type (not just that `K_THREAD_STACK_BUFFER()` will always return a `char *` type) then that is fine. This seems to be the case from your comment below.